File: validBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
//Iterative Solution
var isValidBST = function(root) {
    //iterative
    let prev = null;
    let st = [];
    while(root != null || st.length ){
        while(root != null){
            st.push(root);
            root = root.left;
        }
        root = st.pop();
        if(prev!=null && prev.val >= root.val) return false;
        prev = root;
        root = root.right;
    }
    return true;
    
};

//Time Complexity : O(N)-N: no. of nodes
//space conplexity : O(h)-h: height of the tree


//Recursive
let prev = null;
var isValidBST = function(root) {
    return inorder(root);
    
};

var inorder = function(root) {
    //base
    if(root == null) return true;
    //logic
    inorder(root.left);
    if(prev!= null && prev.val >= root.val ) return false;
    prev = root;
    return inorder(root.right);
}

//time complexity :O(N)
//space complexity : O(h)

File: ValidBST.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    bool result;
    TreeNode prev;
    public bool IsValidBST(TreeNode root) {
        result = true;
        helper(root);
        return result;        
    }

    private void helper(TreeNode root)
    {
        if(root == null || !result) return;
        helper(root.left);
        if(prev != null && prev.val >= root.val)
        {
            result = false;
        }
        prev = root;
        helper(root.right);
    }
}
